ibis/pyspark/tests/test_timecontext.py

```diff
@@ -1,15 +1,227 @@
+    2020-01-04   d
+    with context = (2020-01-03, 2002-01-04) trailing count for 1 day will be:
+    time       value  count
+    2020-01-03   c      2
+    2020-01-04   d      2
+    trailing count for 2 days will be:
+    time       value  count
+    2020-01-03   c      3
+    2020-01-04   d      3
+    with context = (2020-01-01, 2002-01-02) count for 1 day forward looking
+    window will be:
+    time       value  count
+    2020-01-01   a      2
+    2020-01-02   b      2
+    """
+    table = client.table('time_indexed_table')
+    context = (
+        pd.Timestamp('20170102 07:00:00', tz='UTC'),
+        pd.Timestamp('20170103', tz='UTC'),
+    )
+    result_pd = table.mutate(
+        count=table['value'].count().over(ibis_window)
+    ).execute(timecontext=context)
+    spark_table = table.compile()
+    spark_window = (
+        Window.partitionBy('key')
+        .orderBy(F.col('time').cast('long'))
+        .rangeBetween(*spark_range)
+    )
+    expected = spark_table.withColumn(
+        'count', F.count(spark_table['value']).over(spark_window),
+    ).toPandas()
+    expected = expected[
+        expected.time.between(*(t.tz_convert(None) for t in context))
+    ].reset_index(drop=True)
+    tm.assert_frame_equal(result_pd, expected)
+
+
+def test_cumulative_window(client):
+    """ Test context adjustment for cumulative window
+
+    For cumulative window, by defination we should look back infinately.
+    When data is trimmed by time context, we define the limit of looking
+    back is the start time of given time context. Thus for a table of
+    time       value
+    2020-01-01   a
+    2020-01-02   b
+    2020-01-03   c
+    2020-01-04   d
+    with context = (2020-01-02, 2002-01-03) cumulative count will be:
+    time       value  count
+    2020-01-02   b      1
+    2020-01-03   c      2
+    """
+    table = client.table('time_indexed_table')
+    context = (
+        pd.Timestamp('20170102 07:00:00', tz='UTC'),
+        pd.Timestamp('20170105', tz='UTC'),
+    )
+    window = ibis.cumulative_window(order_by='time', group_by='key')
+    result_pd = table.mutate(
+        count_cum=table['value'].count().over(window)
+    ).execute(timecontext=context)
+    df = table.execute()
+    df = df[df.time.between(*(t.tz_convert(None) for t in context))]
+    expected_cum_win = (
+        df.set_index('time')
+        .groupby('key')
+        .value.expanding()
+        .count()
+        .rename('count_cum')
+        .astype(int)
+    )
+    df = df.set_index('time')
+    df = df.assign(
+        count_cum=expected_cum_win.sort_index(
+            level=['time', 'key']
+        ).reset_index(level='key', drop=True)
+    )
+    expected = df.sort_values(by=['key', 'time']).reset_index()
+    tm.assert_frame_equal(result_pd, expected)
+
+
+@pytest.mark.xfail(
+    reason='Issue #2453 chain mutate() for window op and'
+    'non window op throws error for pyspark backend',
+    strict=True,
+)
+def test_complex_window(client):
```

Comment:
I think it's fine to have a complex test, but we should also have simple tests for:
(1) add multiple rolling window ops and verify context is adjusted correctly
(2) add a rolling window op and a cumulative window op
(3) add a rolling window op and a non window op

Reply:
Great idea, I added these tests, but mark some of them xfailed for https://github.com/ibis-project/ibis/issues/2457